### PR TITLE
Slarre reminders app

### PR DESCRIPTION
**Pull request by slarre**

### CODE REVIEW

**Comment by albertaouad** [comment]
Comment from Sebastien by email on pending work:
> I just need to develop a simple frontend in React JS, some minor features (like enable HTTPS), some technical debt and more documentation.
I'm having quite a busy week at work, but I will try to finish by friday.

File: api_gateway_srv/gateway/api/controller.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
+from .microservices_api import user_service, reminder_service
+
+def login_user(username, password):
+    user_srv = user_service()
+
+    # Get user 
+    response = user_srv.get_user(username)
+    if not response:
+        # If doesn't exist => register user
+        if not user_srv.add_user(username, password):
```
Comment (abdoulaye-byd) [inline api_gateway_srv/gateway/api/controller.py]: I'm not sure I understand the design decision behind creating a user on the fly here whenever we try to login with a non-existing user.

File: reminder_srv/dockerfiles/Dockerfile.api [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+COPY ../requirements.txt requirements.txt
+COPY ../app.py app.py
+RUN pip install --upgrade pip
+RUN pip install -r requirements.txt
+RUN pip install gunicorn
+
+COPY ../reminder reminder
+COPY ../boot_api.sh ./
+RUN chmod +x boot_api.sh
```
Comment (abdoulaye-byd) [inline reminder_srv/dockerfiles/Dockerfile.api:15]: Small tip: you can copy files/folders and set the owner directly in one command `COPY --chown {user}:{group} {source} {dest}`

**Comment by abdoulaye-byd** [comment]
Hi @slarre thank you for taking the time to do the assignment.

Here are a couple of things that my highlights on your implementation:

1. I like the architecture diagram, it's a pretty good way to showcase how the overall system work together in one glance.
2. The code is clear, easy to follow and well separated into different directories.
3. Kudos for the OpenApi documentation
4. Nginx reverse proxy and the API gateway to centralize access of the different micro-services
5. Good use of the HTTP status codes in your responses

On the other hand, below are some things that I think could be improved:

1. User registration is automatically happening on the fly during a login when the user does not exist. Though this could just be a simplification because it's a PoC, if that's the case please disregard the comment.
2. From an API design perspective the `reminder` entity seems to be split between the user and reminder service. The list of reminders a user has created are stored within the user service and then the details of the reminder are stored in the reminder service. This creates a very tight coupling between the user and reminder services so that whenever you want to fetch reminders created by a user, you have to first read from the user service and then from the reminder service. Not only this is a performance bottleneck but the system as whole is lot less resilient because if say your user service is down for whatever reason, you can't use your reminder service. A better approach could have been to store and `owner` (the user id) field in the reminder service and directly query the reminder service.
3. This is very subjective, but since you've got an API gateway, you could have standardized the API response from it for better integrity for client consumers of the API using something like `{"data": { }}` and `{"error": { "code": "", "description": ""}` for successful and failed responses respectively.

File: api_gateway_srv/gateway/api/controller.py (hunk 1 above)
Comment (slarre) [reply api_gateway_srv/gateway/api/controller.py]: I know, I didn't like either. It was only to keep UI as simple as possible to solve other things first. If I have time I will add a Sign up UI and I could also add a Change Pwd UI. Anyway User microservice it has both features already implemented. So I should only add proper Api Gateway routes.

**Comment by albertaouad** [comment]
@slarre , once the ReactJS part is in place we can re-review to conclude our evaluation.
Thank you again for taking the time and putting the effort to do this.

**Comment by slarre** [comment]
> From an API design perspective the reminder entity seems to be split between the user and reminder service. The list of reminders a user has created are stored within the user service and then the details of the reminder are stored in the reminder service. This creates a very tight coupling between the user and reminder services so that whenever you want to fetch reminders created by a user, you have to first read from the user service and then from the reminder service. Not only this is a performance bottleneck but the system as whole is lot less resilient because if say your user service is down for whatever reason, you can't use your reminder service. A better approach could have been to store and owner (the user id) field in the reminder service and directly query the reminder service.

I've spend like an hour thinking about this thing I was about to do it in that way. But he reason (probably a wrong one) behind my decision was that conceptually a User owns all his Reminders so, maybe in the future, I could relate with different user data or query some info easily. And left to the reminder service only the possibility to store and manage all kind of reminders data unrelated with anything else. Anyway I understood your concerns and I will remember them in the future.

**Comment by albertaouad** [comment]
@slarre thank you again for taking the time. 

The team completed their review and you will be hearing from us very shortly!

**Pull request merged**